File: 9 - matrizesEspeciais/Celula.java
import java.util.HashSet;

public class Celula {
    int linha;
    int coluna;
    int valor;
    static int qtdLinhas;
    static int qtdColunas;

    public Celula(int linha, int coluna, int valor) {
        this.linha = linha;
        this.coluna = coluna;
        this.valor = valor;
    }

    @Override
    public String toString() {
        return "Celula [linha=" + linha + ", coluna=" + coluna + ", valor=" + valor + "]";
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + linha;
        result = prime * result + coluna;
        return result;
    }

    public static void transformaMatrizEmHash(int matriz[][], HashSet<Celula> tabelaHash) {
        for (int i = 0; i < matriz.length; i++) {
            for (int j = 0; j < matriz[i].length; j++) {
                if (matriz[i][j] != 0) {
                    tabelaHash.add(new Celula(i,j,matriz[i][j]));
                }
            }
        }
        Celula.qtdLinhas = matriz.length;
        Celula.qtdColunas = matriz[0].length;
    }

    public static void transformaHashEmMatriz(HashSet<Celula> tabelaHash, int matriz[][]) {
        for (int i = 0; i < matriz.length; i++) {
            for (int j = 0; j < matriz[i].length; j++) {
                matriz[i][j] = 0;
            }
        }
        
        for (Celula ponto : tabelaHash) {
            matriz[ponto.linha][ponto.coluna] = ponto.valor;
        }
    }

    public static void exibeMatrizEmHash(HashSet<Celula> tabelaHash) {
        System.out.println("Matriz["+Celula.qtdLinhas+"]["+Celula.qtdColunas+"]");

        for (Celula ponto : tabelaHash) {
            System.out.println(ponto);
        }
    }

}

File: 09 - matrizesEspeciais/python/celula.py
class Celula:
    def __init__(self, linha, coluna, valor):
        self.linha = linha
        self.coluna = coluna
        self.valor = valor    

    def __hash__(self):
        return hash((self.linha, self.coluna, self.valor))

    def __repr__(self):
        return f'Celula({self.linha}, {self.coluna}, {self.valor})'

    @staticmethod
    def transforma_matriz_em_hash(matriz, tabela_hash):
        for i in range(len(matriz)):
            for j in range(len(matriz[i])):
                if matriz[i][j] != 0:
                    tabela_hash.add(Celula(i, j, matriz[i][j]))
        
        Celula.qtd_linhas = len(matriz)
        Celula.qtd_colunas = len(matriz[0])

    @staticmethod
    def transforma_hash_em_matriz(tabela_hash, matriz):
        for i in range(len(matriz)):
            for j in range(len(matriz[i])):
                matriz[i][j] = 0
        
        for ponto in tabela_hash:
            matriz[ponto.linha][ponto.coluna] = ponto.valor        

    @staticmethod
    def exibe_matriz_em_hash(tabela_hash):
            print("Matriz[",Celula.qtd_linhas,"][",Celula.qtd_colunas,"]")

            for ponto in tabela_hash:
                print(ponto)
            


